### PR TITLE
[WIP] Restructure io

### LINKED ISSUE #190: Cleanup the namespace and the arguments
While reviewing the module, I noticed that there isn't a clean namespace. In prep for adding pipelines to the module, we should restructure it. 
- We should have the read and write functions live in `mne_bids.io`. this would be either with the function signature there, or having the name called from the io `__init__` file. 
- ~~We have `make_bids_basename` but it does not include the bids_directory base. I think we should include it in the function signature and have the full basename path.~~
- ~~Afterwards we could then remove bids_root from the `read_raw_bids` given that the basename would now contain this information~~

### PR DESCRIPTION
**Pull request by sappelhoff**
closes #190 

WIP --> tests WILL fail, and there are lots of outstanding import issues

### CODE REVIEW

**Comment by sappelhoff** [comment]
first draft for the restructuring:

```
./mne_bids
├── cli
│   ├── __init__.py
│   └── mne_bids_raw_to_bids.py
├── config.py
├── copyfiles.py
├── datasets.py
├── fixes.py
├── __init__.py
├── io
│   ├── base.py
│   ├── __init__.py
│   ├── tests
│   │   ├── test_base.py
│   │   └── test_utils.py
│   └── _utils.py
├── pick.py
├── tests
│   ├── test_cli.py
│   ├── test_copyfiles.py
│   ├── test_datasets.py
│   ├── test_pick.py
│   ├── test_tsv_handler.py
│   └── test_utils.py
├── tsv_handler.py
└── utils.py
```
new
-----
- copyfiles.py and test_copyfiles.py
- the /io directory with base, init, _utils, and a test directory (with tests for base and _utils)

removed / migrated
-------------------------
- io.py and test_io.py
- mne_bids.py and test_mne_bids.py
- utils.py (now only `print_dirtree` is in there ...)

@jasmainak @teonbrooks @massich

File: mne_bids/io/tests/test_base.py [hunk 1]
```diff
@@ -1,15 +1,3 @@
...
-
-For each supported file format, implement a test.
-"""
-# Authors: Mainak Jas <mainak.jas@telecom-paristech.fr>
-#          Teon L Brooks <teon.brooks@gmail.com>
-#          Chris Holdgraf <choldgraf@berkeley.edu>
-#          Stefan Appelhoff <stefan.appelhoff@mailbox.org>
-#          Matt Sanderson <matt.sanderson@mq.edu.au>
-#
-# License: BSD (3-clause)
```
Comment (jasmainak) [inline mne_bids/io/tests/test_base.py]: why did you remove the author list?

**Comment by jasmainak** [comment]
I would like to get @agramfort 's opinion on this PR

**Comment by teonbrooks** [comment]
I like the reorganization. I would move the function in `fixes.py ` and `pick.py` into `utils.py`

File: mne_bids/io/base.py [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+        The first column contains the event time in samples and the third
+        column contains the event id. The second column is ignored for now but
+        typically contains the value of the trigger channel either immediately
+        before the event or immediately after.
+    event_id : dict
+        Dictionary of events in the raw data mapping from the event value
+        to an index.
+
+    """
+    from mne_bids.utils import _parse_bids_filename
```
Comment (agramfort) [inline mne_bids/io/base.py]: nested import should not be necessary

**Comment by agramfort** [comment]
let's discuss this morning

**Comment by sappelhoff** [comment]
> I like the reorganization. I would move the function in fixes.py and pick.py into utils.py

I like that idea. @massich also suggested to move `/cli` outside of the `mne_bids` package, because it is not really part of the code (it's not imported anywhere) ... it's more of a plugin that can sit at the same level of `setup.py`, `docs/`, ...
```

Would it make sense to make a new `/utils` sub-package on the same level as the new `/io` sub-package? In that `/utils` sub-package, we could have
- tsv_handler.py
- copyfiles.py
- utils.py (would be renamed to e.g., print_helpers.py)

And then yet another sub-package `/datasets` with
- datasets.py (but probably put into a structure similar to `/io` with base.py, init.py, and _utils.py)

then at the root we would remain with
- config.py

File: mne_bids/io/__init__.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+"""Initialize IO functions."""
+# Authors: Mainak Jas <mainak.jas@telecom-paristech.fr>
+#          Alexandre Gramfort <alexandre.gramfort@telecom-paristech.fr>
+#          Teon Brooks <teon.brooks@gmail.com>
+#          Chris Holdgraf <choldgraf@berkeley.edu>
+#          Stefan Appelhoff <stefan.appelhoff@mailbox.org>
+#          Matt Sanderson <matt.sanderson@mq.edu.au>
+#
+# License: BSD (3-clause)
+from mne_bids.io.base import (read_raw_bids, write_raw_bids,
```
Comment (jasmainak) [inline mne_bids/io/__init__.py]: this is dangerous. You should always use relative imports. Otherwise, you might end up using the wrong files

File: mne_bids/copyfiles.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    # Copy data .eeg ... no links to repair
+    sh.copyfile(eeg_file_path, fname_dest + '.eeg')
+
+    # Write new header and marker files, fixing the file pointer links
+    # For that, we need to replace an old "basename" with a new one
+    # assuming that all .eeg, .vhdr, .vmrk share one basename
+    __, basename_src = op.split(fname_src)
+    assert basename_src + '.eeg' == op.split(eeg_file_path)[-1]
+    assert basename_src + '.vmrk' == op.split(vmrk_file_path)[-1]
+    __, basename_dest = op.split(fname_dest)
```
Comment (jasmainak) [inline mne_bids/copyfiles.py]: why double underscore?

File: mne_bids/copyfiles.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    # extract encoding from brainvision header file, or default to utf-8
+    enc = _get_brainvision_encoding(vhdr_src, verbose=True)
+
+    # Copy data .eeg ... no links to repair
+    sh.copyfile(eeg_file_path, fname_dest + '.eeg')
+
+    # Write new header and marker files, fixing the file pointer links
+    # For that, we need to replace an old "basename" with a new one
+    # assuming that all .eeg, .vhdr, .vmrk share one basename
+    __, basename_src = op.split(fname_src)
```
Comment (jasmainak) [inline mne_bids/copyfiles.py]: why double underscore?

File: mne_bids/io/__init__.py (hunk 3 above)
Comment (sappelhoff) [reply mne_bids/io/__init__.py]: pep8 to my rescue

https://www.python.org/dev/peps/pep-0008/#imports

> Absolute imports are recommended, as they are usually more readable and tend to be better behaved (or at least give better error messages)

File: mne_bids/io/tests/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+import pytest
+
+from mne.utils import _TempDir
+
+from mne_bids.io._utils import _parse_ext, _read_raw
+
+base_path = op.join(op.dirname(mne.__file__), 'io')
+
+
+
```
Comment (jasmainak) [inline mne_bids/io/tests/test_utils.py]: too many whitespace lines
